proposals/p2240.md

```diff
@@ -189,56 +199,86 @@ fn TupleConcat[T1s: [type;], T2s: [type;]](
+usually marked by the `[:]` operator, but can also be unmarked usages of names
+that have pack types (which will be defined below). The _arity_ of an expansion
+argument is a compile-time value representing the number of elements it
+evaluates to. Every pack expansion must contain at least one expansion argument,
+and all arguments of a given expansion must have the same arity (which we will
+also refer to as the arity of the expansion). If any expansion argument is a
+pattern, the entire pack expansion is a pattern. In particular, this means that
+the expansion arguments of a `...{` expansion must be expressions.
+
+Pack expansions can be nested only if the inner expansion is within one of the
+outer expansion's arguments, although we may relax this restriction in the
+future.
+
+There are two complementary models for the meaning of pack expansions, the
+_syntactic_ model and the _reified_ model. In the syntactic model, we treat a
+pack expansion as being rewritten during monomorphization as a series of copies
+of the pack expansion, where in each copy, each expansion argument is replaced
+with one of its elements. In the reified model, we define the semantics in terms
+of _pack values_, which are created by the `[:]` operator and consumed by
+`...,`, `...and`, `...or`, and `...{`, and we generalize most non-variadic
+operations to apply element-wise to packs.
+
+Both models work well for run-time expression evaluation, and should be
+equivalent in that context. The syntactic model provides a unified description
+of run-time expression and statement semantics, but is difficult to apply to
+symbolic computation, such as type checking and pattern matching. On the other
+hand, the reified model works well for symbolic computation and expression
+evaluation, but does not naturally extend to statement semantics.
+
+We will rely primarily on the reified model, with two exceptions:
+
+-   We will use the syntactic model to describe the run-time semantics of `...{`
+    expansions, because the reified model can't easily describe the semantics of
+    statements, and it's difficult to bridge between the two models within a
+    single pack expansion.
+-   We will sometimes use the syntactic model when describing the run-time
+    semantics of expressions, because expressions inside `...{` expansions must
+    use that model, and because run-time expressions in other expansions will
+    likely be implemented using that model.
+
+Type-checking and pattern-matching will always use the reified model, because
+they are inherently symbolic.
+
+### Syntactic semantics
+
+In the syntactic model, the semantics of packs and pack expansions are specified
+in terms of a procedure for rewriting an AST rooted at `...{`, `...,`, `...and`,
+or `...or` to an equivalent AST that does not. This rewrite takes place at the
+same time as monomorphization of generic functions, which means it takes place
+after name resolution and typechecking.
```

Comment:
It happens after checked generic typechecking, but not as clearly after template typechecking.

It also takes place when the values of symbolic constants are known, which is important to the following.